Exchange rate end point

DusanBrankovic opened the pull request.

DusanBrankovic committed:
Created ExchangeRate class. Created migration file for exchange_rates table.

DusanBrankovic committed:
Created ExchangeRate class. Created migration file for exchange_rates table.

DusanBrankovic committed:
Created ExchangeRate class. Created migration file for exchange_rates table.

DusanBrankovic committed:
Created ExchangeRate class. Created migration file for exchange_rates table.

DusanBrankovic committed:
Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate.

DusanBrankovic committed:
Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate.

DusanBrankovic committed:
Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate.

DusanBrankovic committed:
Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate.

DusanBrankovic committed:
Removed @Unsigned annotation from amountInRsd property in ExchangeRate class. Changed ExchangeRateResponse to match with ExchangeRate.

DusanBrankovic committed:
Removed @Unsigned annotation from amountInRsd property in ExchangeRate class. Changed ExchangeRateResponse to match with ExchangeRate.

DusanBrankovic committed:
Removed @Unsigned annotation from amountInRsd property in ExchangeRate class. Changed ExchangeRateResponse to match with ExchangeRate.

DusanBrankovic committed:
Created request DTO for ExchangeRate.

DusanBrankovic committed:
Created request DTO for ExchangeRate.

DusanBrankovic committed:
Created request DTO for ExchangeRate.

DusanBrankovic committed:
Created request DTO for ExchangeRate.

DusanBrankovic committed:
Created Mapper class for ExchangeRate and its DTOs

DusanBrankovic committed:
Created Mapper class for ExchangeRate and its DTOs

DusanBrankovic committed:
Created Mapper class for ExchangeRate and its DTOs

DusanBrankovic committed:
Created ExchangeRateController with shell endpoints.

DusanBrankovic committed:
Created ExchangeRateController with shell endpoints.

DusanBrankovic committed:
Created ExchangeRateController with shell endpoints.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Created ExchangeRatePutRequest.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Created ExchangeRatePutRequest.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Created ExchangeRatePutRequest.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Created ExchangeRatePutRequest.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Added attributes in ExchangeRatePutRequest.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Added attributes in ExchangeRatePutRequest.

DusanBrankovic committed:
Renamed ExchangeRateRequest to ExchangeRatePostRequest. Added attributes in ExchangeRatePutRequest.

DusanBrankovic committed:
Injected ExchangeRateService in ExchangeRateController.

DusanBrankovic committed:
Injected ExchangeRateService in ExchangeRateController.

DusanBrankovic committed:
Injected ExchangeRateService in ExchangeRateController.

DusanBrankovic committed:
Changed mapExchangeRateRequestToExchangeRate -> mapExchangeRatePostRequestToExchangeRate. Added mapping from ExchangeRatePutRequest to ExchangeRate.

DusanBrankovic committed:
Changed mapExchangeRateRequestToExchangeRate -> mapExchangeRatePostRequestToExchangeRate. Added mapping from ExchangeRatePutRequest to ExchangeRate.

DusanBrankovic committed:
Changed mapExchangeRateRequestToExchangeRate -> mapExchangeRatePostRequestToExchangeRate. Added mapping from ExchangeRatePutRequest to ExchangeRate.

DusanBrankovic committed:
Revert "Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate."

This reverts commit 0dce1a68

DusanBrankovic committed:
Revert "Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate."

This reverts commit 0dce1a68

DusanBrankovic committed:
Revert "Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate."

This reverts commit 0dce1a68

DusanBrankovic committed:
Revert "Changed property names in ExchangeRate class and migration file to more accurately represent what info they hold. Changed hibernate property from update to validate."

This reverts commit 0dce1a68

DusanBrankovic committed:
Removed migration script for ExchangeRate model.

DusanBrankovic committed:
Removed migration script for ExchangeRate model.

DusanBrankovic committed:
Removed migration script for ExchangeRate model.

DusanBrankovic committed:
Removed migration script for ExchangeRate model.

DusanBrankovic committed:
Continued implementation of exchange rate endpoint

* Changes in ExchangeRate:
  - changed `amountInRsd` type from Float to BigDecimal for better control over data
  - renamed `currency` to `code`
  - removed @NotNull annotation from `code` field since @NotBlank covers not null constraint
  - added @Builder, @AllArgsConstructor and @NoArgsConstructor lombok annotations

* Changes in ExchangeRateController:
  - added apiPrefix property to application.properties, and referenced it in @RequestMapping for this endpoint
  - injected ExchangeRateMapper through constructor (it was previously in ExchangeRateServiceLocalImpl)
  - implemented methods for CRUD operations (getAll, getById, create, update, delete)
  - added @Valid annotations to method arguments to check if received data is valid

* Changes in ExchangeRateMapper:
  - added @Component annotation so the class is auto-detected
  - added `id` in mapExchangeRatePutRequestToExchangeRate method signature

* Changes in ExchangeRatePostRequest:
  - added @AllArgsConstructor and @NoArgsConstructor lombok annotations
  - changed `amountInRsd` type from Float to BigDecimal
  - added validation for fields

* Changes in ExchangeRatePutRequest:
  - added @AllArgsConstructor and @NoArgsConstructor lombok annotations
  - changed `amountInRsd` type from Float to BigDecimal
  - added validation for fields

* Changes in ExchangeRateResponse:
  - changed `amountInRsd` type from Float to BigDecimal

+ Created ExchangeRateAlreadyDefined exception
+ Created ExchangeRateNotFound exception
+ Created ExchangeRateRepository
+ Created ExchangeRateService interface
+ Created ExchangeRateServiceLocalImpl:
  - implemented methods for CRUD operations (getAll, getById, create, update, delete)

DusanBrankovic committed:
Continued implementation of exchange rate endpoint

* Changes in ExchangeRate:
  - changed `amountInRsd` type from Float to BigDecimal for better control over data
  - renamed `currency` to `code`
  - removed @NotNull annotation from `code` field since @NotBlank covers not null constraint
  - added @Builder, @AllArgsConstructor and @NoArgsConstructor lombok annotations

* Changes in ExchangeRateController:
  - added apiPrefix property to application.properties, and referenced it in @RequestMapping for this endpoint
  - injected ExchangeRateMapper through constructor (it was previously in ExchangeRateServiceLocalImpl)
  - implemented methods for CRUD operations (getAll, getById, create, update, delete)
  - added @Valid annotations to method arguments to check if received data is valid

* Changes in ExchangeRateMapper:
  - added @Component annotation so the class is auto-detected
  - added `id` in mapExchangeRatePutRequestToExchangeRate method signature

* Changes in ExchangeRatePostRequest:
  - added @AllArgsConstructor and @NoArgsConstructor lombok annotations
  - changed `amountInRsd` type from Float to BigDecimal
  - added validation for fields

* Changes in ExchangeRatePutRequest:
  - added @AllArgsConstructor and @NoArgsConstructor lombok annotations
  - changed `amountInRsd` type from Float to BigDecimal
  - added validation for fields

* Changes in ExchangeRateResponse:
  - changed `amountInRsd` type from Float to BigDecimal

+ Created ExchangeRateAlreadyDefined exception
+ Created ExchangeRateNotFound exception
+ Created ExchangeRateRepository
+ Created ExchangeRateService interface
+ Created ExchangeRateServiceLocalImpl:
  - implemented methods for CRUD operations (getAll, getById, create, update, delete)

DusanBrankovic committed:
Merge remote-tracking branch 'origin/exchange-rate' into exchange-rate

# Conflicts:
#	travel/src/main/java/com/tx/travel/controller/ExchangeRateController.java
#	travel/src/main/java/com/tx/travel/mapper/ExchangeRateMapper.java
#	travel/src/main/java/com/tx/travel/model/ExchangeRate.java
#	travel/src/main/java/com/tx/travel/payload/request/ExchangeRatePostRequest.java
#	travel/src/main/java/com/tx/travel/payload/request/ExchangeRatePutRequest.java
#	travel/src/main/java/com/tx/travel/payload/response/ExchangeRateResponse.java

DusanBrankovic committed:
Removed validateByCode method from interface.

DusanBrankovic committed:
Added methods (validateByCode, validateCodeContent) to validate currency code in ExchangeRateServiceLocalImpl.

Removed validateByCode() method from ExchangeRateService interface.

Renamed ExchangeRateNotFound to ExchangeRateNotFoundException.
Renamed ExchangeRateAlreadyDefined to ExchangeRateAlreadyDefinedException.

Created new ExchangeRateCodeNotValidException.
Created new ExchangeRateAmountInRsdNotValidException.
Created ExchangeRateServiceLocalImplTest.

DusanBrankovic committed:
Merge remote-tracking branch 'origin/exchange-rate' into exchange-rate

# Conflicts:
#	travel/src/main/java/com/tx/travel/controller/ExchangeRateController.java
#	travel/src/main/java/com/tx/travel/model/ExchangeRate.java
#	travel/src/main/java/com/tx/travel/service/ExchangeRateService.java
#	travel/src/main/java/com/tx/travel/service/ExchangeRateServiceLocalImpl.java

DusanBrankovic committed:
Removed unused Exceptions.

megicivovic commented on travel/src/main/java/com/tx/travel/controller/ExchangeRateController.java:
```diff
@@ -0,0 +1,108 @@
+    ExchangeRate exchangeRateById;
+
+    try {
+      exchangeRateById = exchangeRateService.getExchangeRateById(id);
```
maybe use optional in the repository

DusanBrankovic replied:
I'm not sure what you mean by that

I have `Optional<ExchangeRate> findById(Long id)` method in Repository.
When I call `repository.findById(id)` in Service method `findExchangeRateById(id)` I check if `Optional.isPresent()` and move on from there. If Optional is empty I throw Exception and catch it here in this try-catch block in Controller.

megicivovic replied:
all good

megicivovic commented on travel/src/main/java/com/tx/travel/controller/ExchangeRateController.java:
```diff
@@ -0,0 +1,108 @@
+      throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage(), e);
+    }
+
+    return ResponseEntity.status(HttpStatus.NO_CONTENT).body("Exchange rate successfully deleted.");
```
🥇

megicivovic commented on travel/src/main/java/com/tx/travel/service/ExchangeRateServiceLocalImpl.java:
```diff
@@ -0,0 +1,96 @@
+    Optional<ExchangeRate> exchangeRateById = exchangeRateRepository.findById(id);
+
+    if (exchangeRateById.isPresent()) return exchangeRateById.get();
+    else throw new NoSuchElementException();
```
Maybe you could have a custom exception here

DusanBrankovic replied:
I will put ExchangeRateNotFoundException

megicivovic commented on travel/src/main/java/com/tx/travel/service/ExchangeRateServiceLocalImpl.java:
```diff
@@ -0,0 +1,96 @@
+    if (exchangeRateByCode.isPresent()) throw new ExchangeRateAlreadyDefinedException(code);
+  }
+
+  private void validateCodeContent(final String code) throws ExchangeRateCodeNotValidException {
```
amazing work on validation!

DusanBrankovic replied:
Thank you!

megicivovic commented on travel/src/main/java/com/tx/travel/service/ExchangeRateServiceLocalImpl.java:
```diff
@@ -0,0 +1,96 @@
+
+    try {
+      exchangeRateRepository.deleteById(id);
+    } catch (EmptyResultDataAccessException e) {
```
this is ok , but maybe you could have fetched by id first to avoid handling DB exception

DusanBrankovic replied:
I've changed it, thank you for the suggestion!

megicivovic commented on travel/src/test/java/com/tx/travel/service/ExchangeRateServiceLocalImplTest.java:
```diff
@@ -0,0 +1,288 @@
+package com.tx.travel.service;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
```
avoid using star imports, only use specific classes

DusanBrankovic replied:
Noted! Thank you

megicivovic commented on travel/src/test/java/com/tx/travel/service/ExchangeRateServiceLocalImplTest.java:
```diff
@@ -0,0 +1,288 @@
+
+  @Test
+  @DisplayName(
+      "given exchange rate" + "when getting exchange rate by id" + "then return exchange rate")
```
```suggestion
      "given existing exchange rate" 
      + "when getting exchange rate by id"
       + "then return exchange rate")
```

megicivovic commented on travel/src/test/java/com/tx/travel/service/ExchangeRateServiceLocalImplTest.java:
```diff
@@ -0,0 +1,288 @@
+
+  @Test
+  @DisplayName(
+      "given exchange rate"
```
```suggestion
      "given non-existing exchange rate"
```

megicivovic reviewed (commented):
Just some small things, otherwise, really impressive!

DusanBrankovic committed:
Changes from PR

* Changes in ExchangeRateServiceLocalImplTest:
  - Removed star imports from ExchangeRateServiceLocalImplTest.
  - Corrected "given" for getExchangeRateById_success and getExchangeRateById_notFound methods.

* Changes in ExchangeRateServiceLocalImpl:
  - Changed deleteExchangeRate implementation to avoid handling DB exception.
  - Changed getExchangeById method to throw ExchangeRateNotFoundException instead of NoSuchElementException.

* Changes in ExchangeRateServiceLocalImplTest:
  - Changed tests for deleteExchangeRate_success and deleteExchangeRate_exchangeRateNotFound to align with new implementation of deleteExchangeRate method in ExchangeRateServiceLocalImpl.

* Changed caught Exception from NoSuchElementException to ExchangeRateNotFoundException in getExchangeRateById.

* Formatting in ExchangeRate class.

petkocfc reviewed (approved):
Nice work 🥇
